### DISCUSSION TITLE
Dataclasses and VSC

### DISCUSSION THREAD

**Discussion by qzcx** [fvutils/pyvsc#137] [Show and tell]
Related to #135 since it deals with decorators.

The [dataclass decorator](https://docs.python.org/3/library/dataclasses.html) works with vsc.

```
import vsc
from dataclasses import dataclass, field

@vsc.randobj
@dataclass
class data:
  max_value : int = 10
  #This is a bit verbose to initialize here, but could maybe get wrapped up somehow
  #Alternatively could use __post_init__() to define vsc rand members
  randomized_value : vsc.types.rand_bit_t = field(init=False, default_factory=lambda:vsc.rand_bit_t(10))

  @vsc.constraint
  def max_val(self):
    self.randomized_value <= self.max_value

a =  data(max_value=15)
a.randomize()
print(a)
```

Output:
`data(max_value=15, randomized_value=7)`

One thing to note is that this `@dataclass` must come after `@vsc.randobj` or else dataclass decorator is thrown away. @mballance, any comment on this? I'm new to using these decorators in general, so not sure if they should be able to be given in either order in this case. This might apply to other decorators as well so worth understanding.

I'm going to keep investigating combining dataclasses and vsc, but wanted to share.

**Comment by qzcx** [comment]
This is a way to wrap the vsc field declaration up to make it neater

```
def vsc_rand_field(bits):
  return field(init=False, default_factory=lambda:vsc.rand_bit_t(bits))

@vsc.randobj
@dataclass
class data:
  max_value : int = 10
  randomized_value : vsc.types.rand_bit_t = vsc_rand_field(10)

```

**Reply by mballance** [reply]
@qzcx, thanks for sharing these observations! While I'm aware of Python dataclasses, I haven't had much chance to work with them myself. 
On the first issue (decorator ordering), I suspect that the dataclass decorator is inserting the __init__ method that the vsc decorator needs to call to create fields. I think adding a documentation section on using dataclasses with vsc makes sense, and this is a requirement that belongs there.
I'd be very interested to see if we can use the type-hint system to capture the things that 'dataclass' needs to properly create vsc fields, such as field width, rand/non-rand, etc. Not sure it's possible, but would be great if it is.

**Reply by qzcx** [reply]
Ah that makes sense that vsc decorator needs __init__() to be defined first so dataclasses must come first.

They make a good pair since randomized objects are typically just data classes. Nice to get all the boilerplate functions (__init__, __repr__, ect) auto-generated for you. 

Type hinting is something I've been wanting to look at more, but a lot of the support is added in the latest python versions. I'm still on 3.6 in my work environment so it hasn't been something I've figured out. dataclasses itself is only back-supported in 3.6 through pip, so it may not be available to folks with older python versions.
